platform/darwin/src/MGLGeoJSONSource.mm

```diff
@@ -7,25 +7,113 @@
+NSString * const MGLGeoJSONBufferOptionName = @"MGLGeoJSONBufferOptionName";
+NSString * const MGLGeoJSONToleranceOptionName = @"MGLGeoJSONOptionsClusterTolerance";
+
+@interface MGLGeoJSONSource ()
+
+@property (nonatomic, readwrite) NSDictionary *options;
+
+@end
+
 @implementation MGLGeoJSONSource
 
-- (instancetype)initWithSourceIdentifier:(NSString *)sourceIdentifier geoJSONData:(NSData *)data {
-    if (self = [super initWithSourceIdentifier:sourceIdentifier]) {
+- (instancetype)initWithSourceIdentifier:(NSString *)sourceIdentifier geoJSONData:(NSData *)data
+{
+    if (self = [super initWithSourceIdentifier:sourceIdentifier])
+    {
         _geoJSONData = data;
     }
     return self;
 }
 
-- (instancetype)initWithSourceIdentifier:(NSString *)sourceIdentifier URL:(NSURL *)url {
-    if (self = [super initWithSourceIdentifier:sourceIdentifier]) {
+- (instancetype)initWithSourceIdentifier:(NSString *)sourceIdentifier geoJSONData:(NSData *)data options:(NS_DICTIONARY_OF(NSString *, id) *)options
+{
+    if (self = [super initWithSourceIdentifier:sourceIdentifier])
+    {
+        _geoJSONData = data;
+        _options = options;
+    }
+    return self;
+}
+
+- (instancetype)initWithSourceIdentifier:(NSString *)sourceIdentifier URL:(NSURL *)url
+{
+    if (self = [super initWithSourceIdentifier:sourceIdentifier])
+    {
+        _URL = url;
+    }
+    return self;
+}
+
+- (instancetype)initWithSourceIdentifier:(NSString *)sourceIdentifier URL:(NSURL *)url options:(NS_DICTIONARY_OF(NSString *, id) *)options
+{
+    if (self = [super initWithSourceIdentifier:sourceIdentifier])
+    {
         _URL = url;
+        _options = options;
     }
     return self;
 }
 
+- (mbgl::style::GeoJSONOptions)mbgl_geoJSONOptions
+{
+    auto options = mbgl::style::GeoJSONOptions();
+    
+    setOptionForKeyWithValue(MGLGeoJSONMaxZoomLevelOptionName, &options.maxzoom, self.options);
+    setOptionForKeyWithValue(MGLGeoJSONBufferOptionName, &options.buffer, self.options);
+    setOptionForKeyWithValue(MGLGeoJSONToleranceOptionName, &options.tolerance, self.options);
+    setOptionForKeyWithValue(MGLGeoJSONClusterRadiusOptionName, &options.clusterRadius, self.options);
+    setOptionForKeyWithValue(MGLGeoJSONClusterMaximumZoomLevelOptionName, &options.clusterMaxZoom, self.options);
+    setOptionForKeyWithValue(MGLGeoJSONClusterOptionName, &options.cluster, self.options);
+    
+    return options;
+}
+
+template<typename T>
+void setOptionForKeyWithValue(NSString *key, T *value, NSDictionary *options)
+{
+    if (! [options objectForKey:key]) return;
+   
+    id keyValue = [options objectForKey:key];
+    BOOL isKindOfNumber = [keyValue isKindOfClass:[NSNumber class]];
+    
+    // if we are dealing with an NSNumber then determine the selector to extract
+    // its value based on its type and then call that selector, assigning the value
+    // to the passed in mbgl options value
+    if (isKindOfNumber)
```

Comment:
Maybe I'm just too tired but it looks like `ValueEvaluator` (renamed from `PropertyValueEvaluator`) [does the opposite](https://github.com/mapbox/mapbox-gl-native/pull/6049/files#diff-9661184742108500b609e44c1341e16dR6), transforming C++ -> to Foundation. `setOptionForKeyWithValue` in this PR converts from Foundation to C++ just like [NSExpression+ MGLAdditions'  mgl_convertedValueWithValue](https://github.com/mapbox/mapbox-gl-native/pull/6049/files#diff-698c604c23220e922e9df03699e408abR24) that is also in #6049.

 [You mentioned something along these lines few weeks ago](https://github.com/mapbox/mapbox-gl-native/pull/6049#issuecomment-240527548). Maybe you mean we should implement the reverse logic using the visitor pattern so we have a way to go from Foundation to C++. We could use such an `Evaluator` here in that case. In any case, we should definitely dry up the duplication in this PR and #6049 when they both land.

Relatedly https://github.com/mapbox/mapbox-gl-native/pull/6217/commits/4213ab9b1b6f6ab01e53cf5ec995fce1ee267faa removes some of the dynamic weirdness in this PR so the two implementations will be easier to reduce however we decide to do it.